book/src/fork-selection.md

```diff
@@ -0,0 +1,95 @@
+
+
+### Rollback
+
+Before a vote is pushed to the stack, all the votes leading up to vote with a lower lock time than the new vote are popped.  After rollback lockouts are not doubled until the voter catches up to the rollback height of votes.
+
+For example, a vote stack with the following state:
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    4 |         4 |      2  |                    6 |
+|    3 |         3 |      4  |                    7 |
+|    2 |         2 |      8  |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+*Vote 5* is at time 9, and the resulting state is
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    5 |         9 |      2  |                   11 |
+|    2 |         2 |      8  |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+*Vote 6* is at time 10
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    6 |        10 |       2 |                   12 |
+|    5 |         9 |       4 |                   13 |
+|    2 |         2 |       8 |                   10 |
+|    1 |         1 |      16 |                   17 |
+
+At time 10 the new votes caught up to the previous votes.  But *vote 2* expires at 10, so the when *vote 7* at time 11 is applied the votes including and above *vote 2* will be popped.
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:|
+|    7 |        11 |       2 |                   13 |
+|    1 |         1 |      16 |                   17 |
+
+The lockout for vote 1 will not increase from 16 until the stack contains 5 votes.
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
+
+Voters should be rewarded for selecting the fork that the rest of the network selected as often as possible.  This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued.  Thus a reward should be generated for each successful dequeue.
+
+### Economic Finality
+
+Economic finality can be calculated as the total opportunity cost due to the vote lockout at a given fork.  As votes get further and further buried, the economic finality increases because the cost of unrolling would be the total loss of all the reward from the lockouts at that fork.
```

Comment:
I think way more effort needs to go into defining this term.  I don't think of it as something that "increases." There's probably some term closely related to economic finality that increases. Maybe instead of saying "economic finality can be calculated" say something like "<yada yada yada> is called *economic finality*". How about: "a transaction is traditionally labeled *finalized* if a cluster can commit never to reverse it. Solana is an availability system and the cluster naturally breaks into partition as nodes have different views of the active set. Because of this partitioning, finalization is qualified as *economically finalized*. It means the transaction is finalized under the assumption that the fullnode had been following the correct leader. Unrolling transactions to change leaders comes at a cost - all rewards for validating those transactions. That cost is called the *unrolling cost* and can be used to determine *how finalized* a transaction really is. The higher the unrolling cost, the more finalized the transaction."

Reply:
Used in sentences: "A transaction, on average, is confirmed in under 1 second and is 100sol-finalized in under 3 seconds. Unrolling the transaction would cost the partition 100 sols."

Reply:
Incorrect usage, "The economic finality of the transaction is 100 sols."
Correct usage, "The unrolling cost of the transaction is 100 sols."
Correct usage, "Solana implements economic finality, a variation of finality that is qualified by the cost to convince a fullnode to unroll the transaction."
Correct usage, "We'll send your 1 sol transaction once the first transfer is 10sol-finalized. We'll send the 50 sol transaction after it's 100sol-finalized."

Reply:
* "following correct leader" doesn't really make sense.
* network has a synchronous active set, it just doesn't have a synchronized view of the latest fork
> Used in sentences: "A transaction, on average, is confirmed in under 1 second and is 100sol-finalized in under 3 seconds. Unrolling the transaction would cost the partition 100 sols."
It would depend on how many validators voted on fork

Reply:
How about a counter proposal? Some definitions and a set of correct and incorrect usages? The document doesn't define it and then uses it inconsistently. Worse, you replied to one comment above with "in terms of economic finality." What in the world does that mean?

Reply:
```
Finality of *fork A* is defined as the cost in terms of time to the validators                                                                             
to confirm any other fork that does not include *fork A* as an ancestor.     
                                                                                                                                                                                                 
Economic Finality of a fork can be calculatated as the loss of all the    
rewards from *fork A* and its decendants, plus the oportunity cost of reward
due to the exponentially growing lockout. 
```

Reply:
@garious We are using time as the physical bound on finality.   It looks like a combination of 1 and 3, but not exactly.  To actually cause rollback the network still has to spend real time to get to the rollback point.

Reply:
Please, just looking for the answer to the one question. I can't help you otherwise. Is your aim to be consistent with industry precedent or to use the same terms with different definitions?

Reply:
@garious to be consistent with the industry.  Finality means something different in Casper vs Bitcoin, so it has a different meaning for us as well.  But it is something that can be measured in $.  Bitcoin it is cpu/electricity, casper it is stake, for us it is time + stake.

Reply:
For reference, https://vitalik.ca/files/casper_note.html, there's no usage of finality or economic finality that includes a measurement in $. Instead, the definition of economic finality references a different term measured in $.

> This value X is called the *cryptoeconomic security margin* of the finality mechanism.

Reply:
@garious what do you want to call this then?  "rollback difficulty"

Reply:
I suggested "cost of rollback" earlier.

Reply:
@garious The **Economic Finality** of *fork A* can be calculated as the loss of all the                                                                                                                                                                                                                                                                                               rewards from rollback of *fork A* and its descendants, plus the opportunity cost of reward                                                                                                                                                                                                                                                                                                
due to the exponentially growing lockout of the votes that have confirmed *fork A*.